File: src/core/API/operations/TraverseOper.cpp
/*
 Copyright (c) 2012 Petr Koupy
 Copyright (c) 2016 Martin Šícho

 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "core/misc/inout.h"

#include "operations/TraverseOper.hpp"
#include "TraverseOperImpl.hpp"
#include "core/API/ExplorationTreeImpl.h"

TraverseOper::TraverseOper(std::shared_ptr<ExplorationTree> expTree, TraverseCallback& callback) : 
pimpl(new TraverseOper::TraverseOperImpl(expTree, callback)) 
{
    setTreeOperPimpl(pimpl);
}

TraverseOper::TraverseOper(TraverseCallback& callback) : 
pimpl(new TraverseOper::TraverseOperImpl(callback))
{
    setTreeOperPimpl(pimpl);
}

TraverseOper::TraverseOper(
    std::shared_ptr<ExplorationTree> expTree
    , TraverseCallback& callback
    , const std::string& rootSMILES
    ) 
: 
pimpl(new TraverseOper::TraverseOperImpl(expTree, callback, rootSMILES))  
{
    setTreeOperPimpl(pimpl);
}

void TraverseOper::operator()() {
    (*pimpl)();
}


// pimpl

TraverseOper::TraverseOperImpl::TraverseOperImpl(
    std::shared_ptr<ExplorationTree> expTree
    , TraverseCallback& callback
    , const std::string& rootSMILES
) 
:
TreeOperation::TreeOperationImpl::TreeOperationImpl(expTree)
, callback(callback)
, root(expTree->fetchMol(rootSMILES))
{
    // no action
}

TraverseOper::TraverseOperImpl::TraverseOperImpl(
    std::shared_ptr<ExplorationTree> expTree
    , TraverseCallback& callback
) 
:
TreeOperation::TreeOperationImpl::TreeOperationImpl(expTree)
, callback(callback)
{
    // no action
}

TraverseOper::TraverseOperImpl::TraverseOperImpl(TraverseCallback& callback) :
TreeOperation::TreeOperationImpl::TreeOperationImpl()
, callback(callback)
{
    // no action
}

TraverseOper::TraverseOperImpl::TraversalFunctor::TraversalFunctor(
    std::shared_ptr<ExplorationTree> tree
    , TraverseCallback& callback
    ) 
: 
mTree(tree)
, mCallback(callback)
{
    // no action
}

void TraverseOper::TraverseOperImpl::TraversalFunctor::operator()(const std::string& smile, tbb::parallel_do_feeder<std::string>& feeder) const {
    auto tree_pimpl = mTree->pimpl;

    TreeMap::accessor ac;
    tree_pimpl->treeMap.find(ac, smile);
    assert(!ac.empty());

    std::shared_ptr<MolpherMol> mol = ac->second;
//    int x = mol.use_count();
    ac.release();
    makeCallback(mol);
//    SynchCout("use count (before/after): " + parseNumber(x) + "/" + parseNumber(mol.use_count()));

    std::set<std::string>::const_iterator it;
    auto descendants = mol->getDescendants();
    for (it = descendants.begin();
            it != descendants.end(); it++) {
        feeder.add(*it);
    }
}

void TraverseOper::TraverseOperImpl::TraversalFunctor::makeCallback(std::shared_ptr<MolpherMol> morph) const {
    mCallback(morph);
}

void TraverseOper::TraverseOperImpl::operator()() {
    auto tree = getTree();
    if (tree) {
        auto tree_pimpl = tree->pimpl;
        tbb::task_group_context tbbCtx;
        tbb::task_scheduler_init scheduler;
        if (tree_pimpl->threadCnt > 0) {
            scheduler.terminate();
            scheduler.initialize(tree_pimpl->threadCnt);
        }
        ConcurrentSmileVector queue;
        if (!root) {
            root = tree->fetchMol(tree_pimpl->source->getSMILES());
        }
        queue.push_back(root->getSMILES());

        TraversalFunctor functor(tree, callback);
        tbb::parallel_do(queue.begin(), queue.end(), functor, tbbCtx);
    } else {
        throw std::runtime_error("Cannot traverse the tree. None associated with this instance.");
    }
}

File: src/python/molpher/core/operations/TraverseOper.py

# Copyright (c) 2016 Martin Sicho
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import molpher.swig_wrappers.core as core
from molpher.core.operations.TreeOperation import TreeOperation
from molpher.core.operations.callbacks import TraverseCallback
from molpher.core.MolpherMol import MolpherMol

class Callback(TraverseCallback):
    """
    :param callback: the callable to call every time a molecule is encountered during traversal
    :type callback: any callable object with one required parameter

    Basic callback class used to traverse the tree with an arbitrary callable.
    It registers a callable and calls it every time a :term:`morph` is processed.
    It also prevents possible errors and supplies instances of
    `molpher.core.MolpherMol` rather than `molpher.swig_wrappers.core.MolpherMol`.

    """

    def __init__(self, callback):
        super(Callback, self).__init__()
        self._callback = callback

    def __call__(self, morph):
        """
        This method is called by the C++ code. It
        just calls the registered callback for a :term:`morph`

        :param morph: morph in the currently processed tree
        :type morph: `molpher.swig_wrappers.core.MolpherMol`
        """

        morph.__class__ = MolpherMol
        self._callback(morph.tree.fetchMol(
            morph.smiles))  # needs to be done this way (SEGFAULT otherwise, the input parameter seems to be destroyed by SWIG after the call)

class TraverseOper(core.TraverseOper, TreeOperation):

    def __init__(self, tree=None, callback=None, root=None):
        if not callback:
            raise RuntimeError('Callback must be specified.')
        self._callback = Callback(callback)

        if tree:
            super(TraverseOper, self).__init__(tree, self._callback)
        elif tree and root:
            super(TraverseOper, self).__init__(tree, self._callback, root)
        else:
            super(TraverseOper, self).__init__(self._callback)